crates/pessimistic-proof/src/local_state.rs

```diff
@@ -0,0 +1,62 @@
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    certificate::Certificate,
+    local_balance_tree::BalanceTree,
+    local_exit_tree::{hasher::Keccak256Hasher, LocalExitTree},
+    proof::{BalanceRoot, ExitRoot},
+    ProofError,
+};
+
+/// Local state of one network.
+/// The AggLayer tracks the [`LocalNetworkState`] for all networks.
+/// Eventually, this state will be entirely tracked by the networks themselves.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct LocalNetworkState {
+    /// Commitment to the [`BridgeExit`].
+    pub exit_tree: LocalExitTree<Keccak256Hasher>,
+    /// Commitment to the balance for each token.
+    pub balance_tree: BalanceTree,
+}
+
+impl LocalNetworkState {
+    /// Apply the [`Certificate`] on the current [`State`].
+    /// Returns the commitment on the resulting state if successful.
+    pub fn apply_certificate(
+        &mut self,
+        certificate: Certificate,
+    ) -> Result<(ExitRoot, BalanceRoot), ProofError> {
+        // Apply on Exit Tree
+        {
+            let computed_root = self.exit_tree.get_root();
+            if computed_root != certificate.prev_local_exit_root {
+                return Err(ProofError::InvalidLocalExitRoot {
+                    got: computed_root,
+                    expected: certificate.prev_local_exit_root,
+                });
+            }
+
+            for bridge_exit in &certificate.bridge_exits {
+                self.exit_tree.add_leaf(bridge_exit.hash());
+            }
+        }
+
+        // Apply on Balance Tree
+        {
+            for bridge_exit in certificate.bridge_exits {
+                self.balance_tree.withdraw(bridge_exit.token_info, bridge_exit.amount);
+            }
+
+            // Check whether the sender has some debt
+            if self.balance_tree.has_debt() {
+                return Err(ProofError::HasDebt {
+                    network: certificate.origin_network,
+                });
+            }
+        };
```

Comment:
This code block can be reduced to this, I think:

`TokenInfo` can also derive `Copy`

```suggestion
        // Apply on Exit Tree
        let computed_root = self.exit_tree.get_root();
        if computed_root != certificate.prev_local_exit_root {
            return Err(ProofError::InvalidLocalExitRoot {
                got: computed_root,
                expected: certificate.prev_local_exit_root,
            });
        }

        certificate.bridge_exits.iter().for_each(|bridge_exit| {
            self.exit_tree.add_leaf(bridge_exit.hash());
            self.balance_tree.withdraw(bridge_exit.token_info, bridge_exit.amount);
        });

        // Check whether the sender has some debt
        if self.balance_tree.has_debt() {
            return Err(ProofError::HasDebt {
                network: certificate.origin_network,
            });
        }
```